include/core/Concurrency.h

```diff
@@ -207,6 +207,54 @@ CORE_EXPORT
 void noop(double);
 }
 
+//! Run \p functions in parallel using async.
+//!
+//! This executes \p functions on a partition of the indices in the range
+//! [\p start, \p end) using the default async executor.
+template<typename FUNCTION>
+void parallel_for_each(std::size_t start,
+                       std::size_t end,
+                       std::vector<FUNCTION>& functions,
+                       const std::function<void(double)>& recordProgress = concurrency_detail::noop) {
+
+    // Threads access the indices in the following pattern:
+    //   [0, m,   2*m,   ...]
+    //   [1, m+1, 2*m+1, ...]
+    //   [2, m+2, 2*m+2, ...]
+    //   ...
+    // for m partitions. This is choosen because it is expected that the index
+    // will often be used to access elements in a contiguous block of memory.
+    // Provided the threads are doing roughly equal work per call this should
+    // ensure the best possible locality of reference for reads which occur
+    // at a similar time in the different threads.
+
+    std::vector<std::future<bool>> tasks;
+
+    for (std::size_t offset = 0, partitions = functions.size();
+         offset < partitions; ++offset, ++start) {
+
+        // Note there is one copy of g for each thread so capture by reference
+        // is thread safe provided f is thread safe.
+
+        CLoopProgress progress{end - offset, recordProgress,
+                               1.0 / static_cast<double>(partitions)};
+
+        auto& g = functions[offset];
```

Comment:
Historically, this was called `g` because it was in some code which also had an `f` :).

I'm not sure there is a great name for this, it is basically the callable which is the unit of work and I use brief names for these unit of work functions throughout this file. I'm not sure there is convention around this, but I think for these sorts of tasks it would be fairly common to just use a single letter, as in this is some function `f`.

On balance I'd therefore choose to just switch to calling this `f` rather than swap to some different naming throughout this file, if that's okay.

Reply:
I did the rename in [cc7ea3d](https://github.com/elastic/ml-cpp/pull/1277/commits/cc7ea3d7a8fc16e2569b805f2ae7d1d2099a3b76).